File: 74-search-a-2d-matrix/search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int rows = matrix.size();
        int cols = matrix[0].size();

        int t=0, b=rows-1;
        while(t <= b) {
            int midR = (t + b) / 2;
            if(target > matrix[midR][cols - 1]) {
                t = midR + 1;
            } else if(target < matrix[midR][0]){
                b = midR - 1;
            } else {
                break;
            }
        }
        if(t > b) return false;

        int l=0, r=cols-1;
        int midR = (t + b) / 2;
        while(l <= r) {
            int midC = (l + r) / 2;
            if(target > matrix[midR][midC]) {
                l = midC + 1;
            }
            else if(target < matrix[midR][midC]) {
                r = midC - 1;
            }
            else {
                return true;
            }
        }
        return false;
    }   
};

File: 74-search-a-2d-matrix/search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        t, b = 0, len(matrix) - 1
        while t <= b:
            mid_row = (t + b) // 2
            if matrix[mid_row][0] <= target <= matrix[mid_row][-1]:
                break
            if matrix[mid_row][0] > target:
                b = mid_row - 1
            else:
                t = mid_row + 1
        
        if t > b:
            return False
        
        l, r = 0, len(matrix[0]) - 1
        while l <= r:
            mid = (l + r) // 2
            if matrix[mid_row][mid] == target:
                return True
            if matrix[mid_row][mid] > target:
                r = mid - 1
            else:
                l = mid + 1
        return False
            